### PR TITLE
The disposer returned from reaction should be auto bound to correct c…

### PR DESCRIPTION
**Pull request by wangyiz4262**
* [x] Added unit tests
* [x] Added typescript typings
* [x] Verified that there is no significant performance drop (`npm run perf`)

### CODE REVIEW

**Comment by urugator** [comment]
Class instance methods aren't supposed to be destructurable, it doesn't make any sense, it's not a plain object. You could do that with any method and completely kill the point of having prototype (class).

**Comment by wangyiz4262** [comment]
@urugator Thanks for raising your concern, and here are my thoughts:

1. The disposer was not correctly bound before using the arrow function. You have to call `reaction.dispose()` to bind the correct context.
2. With a method defined on prototype gives you the benefit of being able to override from a derived class. As a user, unless you intend to create a child class of `Reaction` and override the dispose method, otherwise I don't see the necessity of keeping the prototype.
3. Your argument of  class instance shouldn't be destructured is very interesting. Could you elaborate more about it?

Thanks!

**Comment by FredyC** [comment]
Personally, I haven't see much of use cases for disposing within the reaction effect callback conditionally. Do you have some practical examples where it can be useful?

I think that explicit call to `reaction.dispose()` might be more sensible so it's apparent at a first glance what is being disposed.

Also, note that moving those methods to instance instead of the prototype might be breaking change for some people who are extending from Reaction (for whatever reasons) or are doing some shenanigans based on content on the prototype.

As such I don't think it's worth it to merge this for a very little gain and potentially big and ugly breaks.

**Comment by urugator** [comment]
`reaction.dispose()`
is an equivalent of 
`Reaction.prototype.dispose.call(reaction)`
There is just a single function in memory - the one defined on prototype.

`dispose = () => {}`
Is an equivalent of 
```javascript
constructor() {
  this.dispose = Reaction.prototype.dispose.bind(this);
}
```
There is a seperate physical copy of that function per every `reaction` instance.

The main benefit you get is memory optimization.
Not every class is inheritable (by design) ... implying that the ability to inherit and override behavior is not the decisive factor for using classes (and therefore for binding it's methods).
In short: If you do use classes, then it doesn't make sense to bind their methods, which unfortunatelly means that you can't pass their methods as standalone references.

**Comment by FredyC** [comment]
So I suppose we shouldn't be merging this. What do you think @urugator?

**Comment by urugator** [comment]
I wouldn't.

**Comment by FredyC** [comment]
@wangyiz4262 Thank you for contribution, but unless you come up with some good use case, it's not worth changing this.
